File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    int dp[10005][205];
    bool solve(vector<int> &nums, int n, int target)
    {
        if(target < 0) return false;
        if(target == 0) return true;
        
        if(dp[target][n] != -1) return dp[target][n];
        
        if(n == 0) return false;
        return dp[target][n] = solve(nums, n-1, target-nums[n-1]) || solve(nums, n-1, target);
    }
    bool canPartition(vector<int>& nums) 
    {
        int s = 0, n = nums.size();
        memset(dp, -1, sizeof(dp));
        
        for(int i = 0; i < n; i++) {
            s += nums[i];
        }
        
        if(s % 2 == 1) return 0;
        return solve(nums, n, s/2);
    }
};

/*

22
11

*/

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    public int F(int[] nums, int s, int n, int[][] dp) {
        if (s == 0 && n == 0) return 1;
        if (s < 0 || n == 0) return 0;

        if (dp[s][n] != -1) return dp[s][n];

        if (nums[n-1] <= s) {
            dp[s][n] = (F(nums, s - nums[n-1], n-1, dp) == 1 || F(nums, s, n-1, dp) == 1) ? 1 : 0;
        } else {
            dp[s][n] = F(nums, s, n-1, dp);
        }
        return dp[s][n];
    }

    public boolean canPartition(int[] nums) {
        int n = nums.length, s = 0;
        for (int ele : nums) {
            s += ele;
        }

        if (s % 2 == 1) return false;

        s = s / 2;
        int[][] dp = new int[s+1][n+1];
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }

        return F(nums, s, n, dp) == 1;
    }
}

/*

1, 5, 5, 11


*/